elastic/kibana
# [Infrastructure UI] New Hosts API

closes: [#152103](https://github.com/elastic/kibana/issues/152103)
fixes: https://github.com/elastic/kibana/issues/151768

## Summary

This PR adds a new API to return host metrics for the Hosts View page. There are some constraints to what Elasticsearch can do in terms of sorting. The main query runs a `terms` aggregation, whose size is dynamic. 

The caveat of this solution is that instead of sorting by the actual metric aggregation, it sorts by either a `min` or `max` aggregation of a field. The [doc](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-terms-aggregation.html#_ordering_by_a_sub_aggregation) mentions that this is the safest way to sort by a sub-aggregation. This could, however, potentially return a wrong order, but due to Elasticsearch limitations, it's much more likely to return wrong results sorting by an `avg` aggregation.

### Comparison

The comparison below was done with a load of 5 days worth of data, containing 500 hosts on each day, on my local env. For short date ranges, Snapshot API performs consistently better, and the new approach performs better with a large amount of data.

#### Snapshot API

Returns all 500 hosts

15 minutes
<img width="317" alt="image" src="https://user-images.githubusercontent.com/2767137/230073865-d0d98ce8-76ff-4295-be9b-94f5b342c9f7.png">

1 day
<img width="317" alt="image" src="https://user-images.githubusercontent.com/2767137/230076350-affbad3d-a498-481e-bcba-bcec019ab739.png">

1 week
<img width="317" alt="image" src="https://user-images.githubusercontent.com/2767137/230077049-320d90c8-4f3e-47ba-bd16-1c5878a83254.png">

<details>
  <summary>Profiler</summary>

**The profiler results was taken running a query with a 1 day date range.**

<img width="1260" alt="image" src="https://user-images.githubusercontent.com/2767137/230080996-af4d1c21-1118-4c17-9c62-d22e362142d8.png">
<img width="1266" alt="image" src="https://user-images.githubusercontent.com/2767137/230081181-2412025f-d3c7-4dd6-ad19-0637e85fc87f.png">
<img width="1266" alt="image" src="https://user-images.githubusercontent.com/2767137/230081244-000adcbb-1abf-4d09-8116-a5d5b9b19050.png">
<img width="1266" alt="image" src="https://user-images.githubusercontent.com/2767137/230081325-95b9fe6d-1382-4fdd-b138-f31611cbd854.png">
</details>


#### Hosts API

Returns up to 100 hosts

15 minutes
<img width="317" alt="image" src="https://user-images.githubusercontent.com/2767137/230073194-cf9b6053-9140-48af-b4b5-0e176285f286.png">

1 day
<img width="317" alt="image" src="https://user-images.githubusercontent.com/2767137/230075902-8c1472c0-b007-4f35-a093-bf6bea2d5066.png">

1 week
<img width="317" alt="image" src="https://user-images.githubusercontent.com/2767137/230076837-580b2a4a-e444-4a2d-b335-ddd19fa9aec2.png">

<details>
  <summary>Profiler</summary>

**The profiler results was taken running a query with a 1 day date range.**

<img width="1266" alt="image" src="https://user-images.githubusercontent.com/2767137/230080856-29ef575f-a031-4072-86e9-e1e66c03a08a.png">

<img width="1266" alt="image" src="https://user-images.githubusercontent.com/2767137/230080733-e0dcd186-e6de-4884-aede-99cdf7fe9e9a.png">
</details>

### How to test

```bash
curl --location -u elastic:changeme 'http://0.0.0.0:5601/ftw/api/metrics/hosts' \
--header 'kbn-xsrf: xxxx' \
--header 'Content-Type: application/json' \
--data '{
    "limit": 100,
    "metrics": [
        {
            "type": "rx"
        },
        {
            "type": "tx"
        },
        {
            "type": "memory"
        },
        {
            "type": "cpu"
        },
        {
            "type": "diskLatency"
        },
        {
            "type": "memoryTotal"
        }
    ],
    "query": {
        "bool": {
            "must": [],
            "filter": [],
            "should": [],
            "must_not": []
        }
    },
    "timeRange": {
        "from": 1680040800000,
        "to":   1680645600000
    },
    "sortField": "rx",
    "sortDirection": "desc",
    "sourceId": "default"
}'
```
